IM peak picking

timosachsenberg opened the pull request:
## Description

This PR introduces a new representation type for ion mobility (IM) data in mass spectra within OpenMS. The new representation is tailored for peak data that has been centroided in the ion mobility dimension. It is designed to handle cases where ion mobility data has undergone centroiding and requires distinct handling compared to other types of IM data.

E.g., if we perform centroid (peak picking) in IM dimension (of mobilograms) we collapse (as discussed with Mohammed) the IM frame into a single spectrum (e.g., sorted by m/z) + a IM float data array.

It now also contains Mohammed's algorithm.


## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

timosachsenberg committed:
propose new IM type

timosachsenberg committed:
Merge branch 'develop' into im_types

jcharkow commented:
Some questions + comments.

1. Is a specific peak picker for ion mobility required or can we just use a current peak picker with different settings? @alhigaylan might know best?
   - An extension of this slightly off topic is would it make sense to just have a single peak picker under MATH maybe even a new folder of PEAK_PICKING or something like that. This will prevent confusing use cases that previously were used like hacking the PeakPickerHiRes to use it on RT peak picking. 

3. I am having some difficulty how the centroided data differs from the concatenated data. 

4. Should the centroiding of IM somehow be tied to m/z centroiding? E.g. would we ever have an instance where IM is centroided and m/z is not centroided?

jcharkow reviewed (commented):
see comments above

timosachsenberg commented:
> Some questions + comments.
> 
> 1. Is a specific peak picker for ion mobility required or can we just use a current peak picker with different settings? @alhigaylan might know best?

@alhigaylan uses same algorithm as PeakPickerHiRes IIRC. I am currently looking into ion mobility trace detection (using the algorithm from MassTraceExtractor). So technically, we can reuse existing low-level algorithms but depending on the data (e.g., MS1 vs. MS2 different algorithms might fit better).
For other developers, it might be beneficial to provide distinct classes (think: similar but not identifcal interfaces) for m/z, chromatogram or IM peak picking. Mainly because different data types require different checks for valid data (e.g., profile data for m/z  picking etc.) and the number of implementations might differ for different data modalities.  

>    * An extension of this slightly off topic is would it make sense to just have a single peak picker under MATH maybe even a new folder of PEAK_PICKING or something like that. This will prevent confusing use cases that previously were used like hacking the PeakPickerHiRes to use it on RT peak picking.

This could be one way. 
E.g. we could think about having a PeakPickerProfileSpectrum, PeakPickerChromatogram, PeakPickerIMFrame classes in such a folder and move the generic code (e.g. template) from PeakPickerHiRes to PeakPicker class. Or as in the screenshot: in the CENTROIDING folder. This might be cleaner would change some python bindings etc..

> 2. I am having some difficulty how the centroided data differs from the concatenated data.

I am also not 100% sure (ping @cbielow) but my impression was that CONCATENATED is always the full frame stored in a single spectrum via concatenation of individual scans. The concatenated data is much larger than the IM centroided one because IM values in the float data array are repeated for all peaks of a scan. 
After IM centroiding, each resulting m/z + IM peak can have different IM values and is usually sorted by m/z (I think this is not true for CONCATENATED as basically scans are just added one after the other starting at low m/z values again). 

> 3. Should the centroiding of IM somehow be tied to m/z centroiding? E.g. would we ever have an instance where IM is centroided and m/z is not centroided?

Excellent point, and I was also wondering about this today. I think we could make a decision here and assume that we only pick in IM dimension if m/z dimension is already picked.

timosachsenberg commented:
Or we just add it to PeakPickerHiRes and add a a parameter to also pick the IM dimension. Less composable but maybe easier for users?

alhigaylan commented:
Thank you @timosachsenberg for starting this up. 

Here is my take for some the points discussed here.


> > Some questions + comments.
> > 
> > 1. Is a specific peak picker for ion mobility required or can we just use a current peak picker with different settings? @alhigaylan might know best?
> 
> @alhigaylan uses same algorithm as PeakPickerHiRes IIRC. I am currently looking into ion mobility trace detection (using the algorithm from MassTraceExtractor). So technically, we can reuse existing low-level algorithms but depending on the data (e.g., MS1 vs. MS2 different algorithms might fit better). For other developers, it might be beneficial to provide distinct classes (think: similar but not identifcal interfaces) for m/z, chromatogram or IM peak picking. Mainly because different data types require different checks for valid data (e.g., profile data for m/z picking etc.) and the number of implementations might differ for different data modalities.
> 

I agree that PeakPickerHiRes is general enough to be used for ion mobility data. Still, we need to do some pre-processing (for example, resampling and smoothing) to make it applicable for HiRes. 


I also want to flag that people may have different objectives for peak picking. My understanding of centroiding is to retain as much data as possible but in a reduced useful format. For instance, what do you do when a mobilogram only contains 2 raw peaks? I believe in the context of centroiding, those peaks should be retained because the overall objective is more downstream analysis (mass trace detection across RT, peptide feature detection, etc). But if a peak picker is to be used for quantification or detection such as the context in OpenSWATH, do we want that hypothetical mobilogram with only 2 raw peaks to be picked? 

Since the difference boils down to parameter settings, we can explain this in the documentation for 'PeakPickerIM'. Maybe the PeakPickerIM default parameters should be strict and pick 'good' looking peaks. I think we should have a separate centroiding algorithm that uses 'PeakPickerIM' internally with the setting that allows it to pick nearly all peaks. 

My current mobilogram pre-processing setup is linear resampling followed by smoothing with Gaussian. This ensures peakpickerHiRes retains patchy peaks. 



> > 2. I am having some difficulty how the centroided data differs from the concatenated data.
> 
> I am also not 100% sure (ping @cbielow) but my impression was that CONCATENATED is always the full frame stored in a single spectrum via concatenation of individual scans. The concatenated data is much larger than the IM centroided one because IM values in the float data array are repeated for all peaks of a scan. After IM centroiding, each resulting m/z + IM peak can have different IM values and is usually sorted by m/z (I think this is not true for CONCATENATED as basically scans are just added one after the other starting at low m/z values again).
> 
 
 It seems like "concatenated" is how diapysef / msconvert writes PASEF data in mzML format. A single 'MSSpectrum' object is 1 frame whose all ion mobility scans are written as one array sorted by m/z (and ion mobility stored in FloatDataArrays()) 
 
> > 3. Should the centroiding of IM somehow be tied to m/z centroiding? E.g. would we ever have an instance where IM is centroided and m/z is not centroided?
> 
> Excellent point, and I was also wondering about this today. I think we could make a decision here and assume that we only pick in IM dimension if m/z dimension is already picked.

 In my approach, the projection of all peaks in one PASEF frame to the m/z dimension addresses this problem. In this step, we merge all ion mobility scans into one array to generate a spectrum that exhibits a raw profile peak shape. Unlike the 'concatenated' PASEF frame, we sum up the intensity of repeated m/z peaks (repeated because they are recorded at different ion mobility values). PeakPickerHiRes identifies the m/z centroid from this step. 
Subsequently, each centroided m/z peak will have its mobilogram extracted and peak picked. Based on the picked ion mobility peak FWHM, we can revisit the raw data and compute intensity weighted m/z centroid based on its ion mobility boundaries. This is relevant in cases where one m/z centroid ends up with two or even three mobilogram peaks.

jcharkow commented:
> > * An extension of this slightly off topic is would it make sense to just have a single peak picker under MATH maybe even a new folder of PEAK_PICKING or something like that. This will prevent confusing use cases that previously were used like hacking the PeakPickerHiRes to use it on RT peak picking.
> 
> This could be one way. E.g. we could think about having a PeakPickerProfileSpectrum, PeakPickerChromatogram, PeakPickerIMFrame classes in such a folder and move the generic code (e.g. template) from PeakPickerHiRes to PeakPicker class. Or as in the screenshot: in the CENTROIDING folder. This might be cleaner would change some python bindings etc..


I like this idea

jcharkow commented:
> I am having some difficulty how the centroided data differs from the concatenated data.
>> I am also not 100% sure (ping @cbielow) but my impression was that CONCATENATED is always the full frame stored in a single spectrum via concatenation of individual scans. The concatenated data is much larger than the IM centroided one because IM values in the float data array are repeated for all peaks of a scan. After IM centroiding, each resulting m/z + IM peak can have different IM values and is usually sorted by m/z (I think this is not true for CONCATENATED as basically scans are just added one after the other starting at low m/z values again).

> It seems like "concatenated" is how diapysef / msconvert writes PASEF data in mzML format. A single 'MSSpectrum' object is 1 frame whose all ion mobility scans are written as one array sorted by m/z (and ion mobility stored in FloatDataArrays())

Ok makes sense seems that the centroided in a way would be processed concatenated data.

timosachsenberg commented:
- I will today add some class/file stubs to this PR for PeakPickerIM and tests.
- We still need to find a reliable way to determine if data is in centroided format. Easiest would be if we assume that we always do the IM picking so we could just annotate it at the spectrum level. But: if we load mzMLs we don't have this information available, and I did not find a CV term in the PSI-MS ontology to annotate this. In this case we would need to determine it based on the data itself (e.g., if a thid-party tool did the IM centroiding) or based on a custom meta value stored by us in the m zML. Then we could do something similar to peaktypeestimator by looking for either the meta value or if missing at the peak patterns in the data arrays.

timosachsenberg committed:
add stub for PeakPickerIM

timosachsenberg commented:
@alhigaylan I added a stub for discussion. e.g. there you could also add your code as a separate static member function

timosachsenberg committed:
make it compile

timosachsenberg committed:
todo for IMTypes

timosachsenberg committed:
add UNKNOWN to IMFormat

timosachsenberg committed:
allow both drift time and IM arrays

timosachsenberg committed:
add to pxd

timosachsenberg commented:
Draft is ready from my side and open for comments / suggestions.

cbielow commented on src/openms/source/METADATA/SpectrumSettings.cpp:
```diff
@@ -21,6 +21,7 @@ namespace OpenMS
   SpectrumSettings::SpectrumSettings() :
     MetaInfoInterface(),
     type_(UNKNOWN),
+    im_type_(IMFormat::UNKNOWN),
```
can you switch to C++ 11 and assign that in the header?
Then default the whole Ctor.

cbielow commented on src/openms/include/OpenMS/METADATA/SpectrumSettings.h:
```diff
@@ -142,6 +151,7 @@ namespace OpenMS
 protected:
 
     SpectrumType type_;
+    IMFormat im_type_;
```
SpectrumSettings is already a whopping 656 bytes per spectrum, i.e. worth about 54 peaks. We should do something about that, but these 4 bytes won't make a difference.

cbielow commented on src/openms/source/IONMOBILITY/IMTypes.cpp:
```diff
@@ -73,33 +73,45 @@ namespace OpenMS
+    if (occs.size() == 1) 
     {
-      return *occs.begin();
+      auto format = *occs.begin();
+      if(!(format != IMFormat::CONCATENATED 
+        || format != IMFormat::MULTIPLE_SPECTRA
+        || format != IMFormat::CENTROIDED))
+        {
+          throw Exception::InvalidValue(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "subfunction returned invalid value(s)", "Number of different values: " + String(occs.size()));
+        }
```
not the same thing as before. It would throw if 3 values were returned, for example.

timosachsenberg replied:
The more serious issue I see is that we don't distinguish for the differetn MS level. But isn't it now better  (assuming that e.g. different level might have different representation and thus more than two formats would be possible?)

cbielow replied:
this
```
if(!(format != IMFormat::CONCATENATED 
        || format != IMFormat::MULTIPLE_SPECTRA
        || format != IMFormat::CENTROIDED))
```
will never be true, for any input.
This basically says: `! (a || b || c)` <> ` (!a & !b & !c)` which means
``` 
if (format == IMFormat::CONCATENATED 
       && format == IMFormat::MULTIPLE_SPECTRA
        && format == IMFormat::CENTROIDED)
```
You meant 
```
if (format != IMFormat::CONCATENATED 
       && format != IMFormat::MULTIPLE_SPECTRA
        && format != IMFormat::CENTROIDED)
``` 
???

timosachsenberg replied:
aaah ok I got it. typo. thanks

cbielow replied:
``` "Number of different values: " + String(occs.size()));``  this will also always be 1.

But yes, if MS levels have different format, just returning `mixed` is probably a good thing.

cbielow replied:
Can you add a test ?

cbielow commented on src/openms/include/OpenMS/METADATA/SpectrumSettings.h:
```diff
@@ -78,6 +79,14 @@ namespace OpenMS
+    /// @param im_type
+    void setIMFormat(const IMFormat& im_type);
+
+    /// @brief returns the IMFormat of the spectrum
+    /// @return IMFormat of the spectrum
+    IMFormat getIMFormat() const;
```
not entirely happy here, since the user may not be aware that the value is not set by default and needs to be computed using IMTypes. Maybe add a comment

timosachsenberg replied:
true but maybe we can set this automatically in the common FileHandlers. will add a comment

cbielow commented on src/openms/include/OpenMS/IONMOBILITY/IMTypes.h:
```diff
@@ -46,6 +46,8 @@ namespace OpenMS
     CONCATENATED,    ///< ion mobility frame is stacked in a single spectrum (i.e. has an IM float data array)
     MULTIPLE_SPECTRA,///< ion mobility is recorded as multiple spectra per frame (i.e. has one IM annotation per spectrum)
     MIXED,           ///< an MSExperiment contains both CONCATENATED and MULTIPLE_SPECTRA
+    CENTROIDED,      ///< ion mobility of peaks after centroiding in IM dimension. Ion mobility is annotated in a single float data array (i.e., each peak might have a different IM value in the data array)
```
```suggestion
    CENTROIDED,      ///< ion mobility of peaks after centroiding in IM dimension. Ion mobility is annotated in a single float data array (i.e., each peak might have a different IM value in the data array); identical to CONCATENATED in terms of data layout.
```

cbielow reviewed (changes requested):
good way forward.
Minor comments.

Putting PPIM into PP directory is fine IMHO. I would look for it there.
We could also add an alias to the IM folder (just a header which includes the other one).

timosachsenberg committed:
apply reviewer suggestions

timosachsenberg committed:
fix logic

timosachsenberg committed:
fix logic add test

timosachsenberg committed:
fix logic add test

timosachsenberg committed:
check for MIXED already in setIMFormat

timosachsenberg committed:
fix test

timosachsenberg committed:
update tests

timosachsenberg committed:
fix removed constructor

alhigaylan commented on src/openms/include/OpenMS/PROCESSING/CENTROIDING/PeakPickerIM.h:
This will be a copy of `PeakPickerHiRes` but with `LinearResampler` and `NoiseFilterGaussian` /  `NoiseFilterSGolay`

jcharkow commented on src/openms/source/PROCESSING/CENTROIDING/PeakPickerIM.cpp:
```diff
@@ -0,0 +1,34 @@
+    {        
+        // determine IM format of spectrum
+        IMFormat format = IMTypes::determineIMFormat(spectrum);
+        switch (format)
```
Should there be IMFormat::Unknown as well for consistency?

timosachsenberg replied:
The result of determineIMFormat is currently never UNKNOWN but I could  handle it in the switch as well to make this more explicit.

jcharkow reviewed (approved):
Looks good to me! Just one minor question

timosachsenberg committed:
add some comments

timosachsenberg committed:
Merge branch 'develop' into im_types

timosachsenberg committed:
Merge branch 'develop' into im_types

Commit:
started to populate peakpickerIM with code

Commit:
included linear resampler

Commit:
added topp tool to feed in real mobilograms

Commit:
PeakPickerIM basic skeleton is complete. It accepts raw TimsTOF frame and output centroided mz and im peaks

Commit:
safety check in case SumFrame returns spectrum.size()==1

Commit:
Merge remote-tracking branch 'origin/develop' into im_types

Commit:
add low memory processing

timosachsenberg commented on src/topp/PeakPickerIM.cpp:
```diff
@@ -30,21 +30,76 @@ class TOPPPeakPickerIM : public TOPPBase
+    String process_option = getStringOption_("processOption");
+
+    PeakPickerIM picker;
+    if (process_option == "lowmemory")
```
this option is new

cbielow replied:
nice

timosachsenberg commented on src/openms/source/PROCESSING/CENTROIDING/PeakPickerIM.cpp:
```diff
@@ -706,6 +708,13 @@ namespace OpenMS
 
+
+      // TODO: check why there are so many mobilogram_traces that are empty. leads to segfault later
+      mobilogram_traces.erase(
+        std::remove_if(mobilogram_traces.begin(), mobilogram_traces.end(),
+                   [](const auto& trace) { return trace.empty(); }),
+        mobilogram_traces.end());
+
```
removing empty traces was somehow necessary. Otherwise I got a segfault on data from Josh

Commit:
disable some debug output

timosachsenberg committed:
Refactor PeakPickerIM: Update SumFrame function signature and improve spectrum processing logging

timosachsenberg committed:
more speed improvements (#7967)

Co-authored-by: timosachsenberg <sachsenb@ibminode05.Cs.Uni-Tuebingen.De>

timosachsenberg commented:
![image](https://github.com/user-attachments/assets/2884d3c0-cc64-45b7-a002-e4e6908aa076)
parallelized

timosachsenberg commented:
Ok wow. Previously I stopped it after 30h. Now finishes in 13 min (80 threads + new optimizations).

cbielow commented on src/tests/class_tests/openms/source/IMTypes_test.cpp:
```diff
@@ -126,14 +126,13 @@ START_SECTION(static IMFormat determineIMFormat(const MSExperiment& exp))
     exp.addSpectrum(IMwithDrift);
     exp.addSpectrum(IMwithFDA);
     exp.addSpectrum(IMwithFDA2);
-    TEST_EXCEPTION(Exception::InvalidValue, IMTypes::determineIMFormat(exp))
```
What should the IMFormat of this spectrum be?
Do we need a new ENUM value?

cbielow commented on src/tests/class_tests/openms/source/IMTypes_test.cpp:
```diff
@@ -147,11 +146,9 @@ START_SECTION(static IMFormat determineIMFormat(const MSSpectrum& spec))
    TEST_EQUAL(IMTypes::determineIMFormat(IMwithFDA) == IMFormat::CONCATENATED, true)
 
-   // set both ... invalid!
+   // set both ... is valid (typically concatenated + some average value)
```
same here

cbielow commented on src/topp/PeakPickerIM.cpp:
```diff
@@ -0,0 +1,127 @@
+    setValidFormats_("out", { "mzML" });
+
+    registerStringOption_("processOption", "<name>", "inmemory", "Whether to load all data and process them in-memory or whether to process the data on the fly (lowmemory) without loading the whole file into memory first", false, true);
+    setValidStrings_("processOption", ListUtils::create<String>("inmemory,lowmemory"));
```
```suggestion
    setValidStrings_("processOption", { "inmemory", "lowmemory" } );
```

cbielow committed:
Update src/topp/PeakPickerIM.cpp

timosachsenberg committed:
Refactors private methods for consistent naming

Renames methods to adhere to private naming conventions by appending underscores. Updates all references accordingly, ensuring consistency across header and implementation files.

Improves code readability and maintains convention adherence.

timosachsenberg committed:
Merge remote-tracking branch 'origin/im_types' into im_types

timosachsenberg committed:
preserve more meta data

Commit:
fix typo

timosachsenberg commented:
I did a quick experiment what happens if we run it on a larger dataset and then feed the output (with picked IM) into OpenSWATH.

We now get:
```
Progress of 'Retention time normalization':
# mz regression parameters: Y = -8.64506 + 0.00952783 X + -3.77567e-06 X^2
 sum residual sq ppm before 186216 / after 116750
# im regression parameters: Y = -0.0183206 + 1.01348 X + 0 X^2
-- done [took 20:08 m (CPU), 31.04 s (Wall)] --
Will analyze 1186380 transitions in total.
Progress of 'Extracting and scoring transitions':
Setting up nested loop with 1 threads out of 40

...
...

Thread 0_21 will analyze 99678 compounds and 598068 transitions from SWATH 2 (batch 33 out of 398)
 This should never happen, pr_it has advanced past the master container: 0.93191  / 0.93169

---------------------------------------------------
FATAL: uncaught exception!
---------------------------------------------------
last entry in the exception handler:
exception of type OutOfRange occurred in line 100, function static void OpenMS::IonMobilityScoring::alignToGrid_(const OpenMS::Mobilogram&, const std::vector<double>&, OpenMS::Mobilogram&, double, OpenMS::Size&) of /beegfs/HPCscratch/sachsenb/OpenMS/src/openms/source/ANALYSIS/OPENSWATH/IonMobilityScoring.cpp
error message: the argument was not in range
---------------------------------------------------
..//./scripts/run_osw.sh: line 183: 341508 Aborted             
```
Any ideas?
- Maybe sorting?
- Maybe the extra float data arrays that are not expected (checked, not the case)
```
Meta data array:
  IM FWHM:      4 spectra
  Ion Mobility: 4 spectra
  MZ FWHM:      4 spectra
```

timosachsenberg commented:
Issue seems to be how  IonMobilityScoring::alignToGrid_ is expecting the data.
If I disable ion mobility scores and it runs through!

Seems to be quite fast now. 
**with the more conservative peak picking by Mohammed:**
IM peak picking ~ **12** min
OSW workflow with library free ~ **21** min (was 2h20 min)

but less sensitive (as it is mainly optimized for MS1 right now):
Total peakgroups (q-value<0.01): 2151 (vs **3946**)
Total peakgroups (q-value<0.05): 2851 (vs **4961**)

XGBoost failed with Error: The estimated pi0 <= 0. Check that you have valid p-values or use a different range of lambda. Current lambda range: [0.1  0.15 0.2  0.25 0.3  0.35 0.4  0.45]

**with simple peak clustering:**
IM peak picking ~ **12** min
OSW workflow with library free ~ **41** min (was 2h20 min)

LDA:
Total peakgroups (q-value<0.01): 3159  (vs **3946**)
Total peakgroups (q-value<0.05): 4057  (vs **4961**)

XGBoost:
Total peakgroups per run (q-value<0.01):  **4337** (vs 4321)
Total peakgroups (q-value<0.05): **5666** (vs 5293)


**with trace detection:**
IM peak picking ~ **5** min
LDA:
Total peakgroups (q-value<0.01): 2199
Total peakgroups (q-value<0.05): 3000

XGBoost:
Total peakgroups (q-value<0.01): 5766
Total peakgroups (q-value<0.05): 7814

XGBoost numbers are a bit too good so take this with a grain of salt

Commit:
add stubs for peak centric algorithms

timosachsenberg committed:
more spectrum centric code

Commit:
fix bug related to mass trace detection only compatible with ms1

Commit:
add other algorithms

Commit:
new cluster algo

timosachsenberg committed:
fix edge case for single traces

timosachsenberg commented:
Tested EPD. Those were sensitive parameters.
Will implement in PeakPickerIM soon
![image](https://github.com/user-attachments/assets/aa5350ad-02dc-4e68-9b42-a01078988f7c)
![image](https://github.com/user-attachments/assets/4448d0ad-396f-43f5-b46f-4df3694f5f52)

RT axis = IM

timosachsenberg committed:
add lossless elution peak detection

timosachsenberg committed:
Adds dependency for elution peak detection

Includes the ElutionPeakDetection header to enable its usage
for feature detection processes within the PeakPickerIM module.

timosachsenberg committed:
Adds support for PeakPickerIM and improves elution peak detection

Introduces PeakPickerIM to the tools map for centroiding operations.
Enhances elution peak detection by handling single mass traces,
adjusting parameters, and improving peak detection logic.
Updates test output to include new parameters like `FWHM_mz_avg`.

Improves code readability and consistency in parameter names and values.

timosachsenberg commented:
superseeded by new PR #8177
